builder/builder.go

```diff
@@ -412,97 +426,448 @@ func (b *Builder) getUpstreamBundles(ver string, prune bool) error {
+	// Fetch upstream bundle files if needed
 	if err := b.getUpstreamBundles(b.UpstreamVer, true); err != nil {
-		return bundleAddCount, err
+		return err
 	}
 
-	// Add all bundles to list if --all passed
-	if allbundles {
-		files, err := ioutil.ReadDir(clrbundledir)
+	// Read in current mix bundles list
+	bundleSet, err := b.getMixBundlesListAsSet()
+	if err != nil {
+		return err
+	}
+
+	// Check that the ones passed in exist
+	for _, bundle := range bundles {
+		path, err := b.getBundlePath(bundle)
 		if err != nil {
-			return bundleAddCount, err
+			return err
 		}
+		if b.isLocalBundle(path) {
+			fmt.Printf("Adding bundle %s from local bundles\n", bundle)
+		} else {
+			fmt.Printf("Adding bundle %s from upstream bundles\n", bundle)
+		}
+	}
 
-		bundles = make([]string, len(files))
+	// Add all local bundles to the bundles
+	if allLocal {
+		files, err := ioutil.ReadDir(b.LocalBundleDir)
+		if err != nil {
+			return errors.Wrapf(err, "Failed to read local bundles dir: %s", b.LocalBundleDir)
+		}
 
-		for i, file := range files {
-			bundles[i] = file.Name()
+		for _, file := range files {
+			bundle := file.Name()
+			bundles = append(bundles, bundle)
+			fmt.Printf("Adding bundle %s from local bundles\n", bundle)
 		}
 	}
 
-	var includes []string
+	// Add all upstream bundles to the bundles
+	if allUpstream {
+		clrbundledir := getUpstreamBundlesPath(b.UpstreamVer)
+		files, err := ioutil.ReadDir(clrbundledir)
+		if err != nil {
+			return errors.Wrapf(err, "Failed to read upstream bundles dir: %s", clrbundledir)
+		}
+
+		for _, file := range files {
+			bundle := file.Name()
+			bundles = append(bundles, bundle)
+			fmt.Printf("Adding bundle %s from upstream bundles\n", bundle)
+		}
+	}
+
+	// Add bundles to set
 	for _, bundle := range bundles {
-		// Check if bundle exists in clrbundledir
-		if _, err := os.Stat(filepath.Join(clrbundledir, bundle)); os.IsNotExist(err) {
-			return bundleAddCount, errors.Errorf("Bundle %s does not exist in CLR version %s", bundle, b.UpstreamVer)
-		}
-		// Check if bundle exists in mix bundles dir
-		if _, err := os.Stat(filepath.Join(b.BundleDir, bundle)); os.IsNotExist(err) || force {
-			if !allbundles {
-				var ib []string
-				// Parse bundle to get all includes
-				if ib, err = helpers.GetIncludedBundles(filepath.Join(clrbundledir, bundle)); err != nil {
-					return bundleAddCount, errors.Errorf("Cannot parse bundle %s from CLR version %s", bundle, b.UpstreamVer)
-				} else if len(ib) > 0 {
-					includes = append(includes, ib...)
-				}
-			}
+		bundleSet[bundle] = true
+	}
 
-			fmt.Printf("Adding bundle %q\n", bundle)
-			if err = helpers.CopyFile(filepath.Join(b.BundleDir, bundle), filepath.Join(clrbundledir, bundle)); err != nil {
-				return bundleAddCount, err
-			}
-			bundleAddCount++
+	// Write final mix bundle list back to file
+	if err := b.writeMixBundleList(bundleSet); err != nil {
```

Comment:
I didn't support this deliberately, because I want this file (and `upstreamversion`, `mixversion`, and `upstreamurl`) to be merged in to `builder.conf`, though that cannot happen until it is a proper TOML file that supports lists.

If I intended this to stay in this form forever, yeah, I would add in better parsing to support skipping comment lines, etc.  And at that point, I might consider a duplicate bundle an error, rather than just eliminating it.

Reply:
Ah! That makes sense.

Optional: add a note for people like me to know that "hey, this will eventually change to be part of that".